examples/llm_plus_gnn/g_retriever.py

```diff
@@ -0,0 +1,472 @@
+    df = pd.concat([pd.DataFrame(d) for d in eval_output])
+    all_hit = []
+    all_precision = []
+    all_recall = []
+    all_f1 = []
+
+    for pred, label in zip(df.pred.tolist(), df.label.tolist()):
+        try:
+            pred = pred.split('[/s]')[0].strip().split('|')
+            hit = re.findall(pred[0], label)
+            all_hit.append(len(hit) > 0)
+
+            label = label.split('|')
+            matches = set(pred).intersection(set(label))
+            precision = len(matches) / len(set(label))
+            recall = len(matches) / len(set(pred))
+            if recall + precision == 0:
+                f1 = 0
+            else:
+                f1 = 2 * precision * recall / (precision + recall)
+
+            all_precision.append(precision)
+            all_recall.append(recall)
+            all_f1.append(f1)
+
+        except Exception as e:
+            print(f'Label: {label}')
+            print(f'Pred: {pred}')
+            print(f'Exception: {e}')
+            print('------------------')
+    hit = sum(all_hit) / len(all_hit)
+    precision = sum(all_precision) / len(all_precision)
+    recall = sum(all_recall) / len(all_recall)
+    f1 = sum(all_f1) / len(all_f1)
+
+    print(f'Hit: {hit:.4f}')
+    print(f'Precision: {precision:.4f}')
+    print(f'Recall: {recall:.4f}')
+    print(f'F1: {f1:.4f}')
+
+    return hit
+
+
+def save_params_dict(model, save_path):
+    state_dict = model.state_dict()
+    param_grad_dict = {
+        k: v.requires_grad
+        for (k, v) in model.named_parameters()
+    }
+    for k in list(state_dict.keys()):
+        if k in param_grad_dict.keys() and not param_grad_dict[k]:
+            # delete parameters that do not require gradient
+            del state_dict[k]
+    torch.save(state_dict, save_path)
+
+
+def load_params_dict(model, save_path):
+    state_dict = torch.load(save_path)
+    model.load_state_dict(state_dict)
+    return model
+
+
+def get_loss(model, batch, model_save_name) -> torch.Tensor:
+    if model_save_name == "llm":
+        return model(batch.question, batch.label, batch.desc)
+    else:
+        return model(batch.question, batch.x, batch.edge_index, batch.batch,
+                     batch.ptr, batch.label, batch.edge_attr, batch.desc)
+
+
+def inference_step(model, batch, model_save_name):
+    if model_save_name == "llm":
+        return model.inference(batch.question, batch.desc)
+    else:
+        return model.inference(batch.question, batch.x, batch.edge_index,
+                               batch.batch, batch.ptr, batch.edge_attr,
+                               batch.desc)
+
+
+def train(since, num_epochs, hidden_channels, num_gnn_layers, batch_size,
+          eval_batch_size, lr, loss_fn, inference_fn, model=None, dataset=None,
+          checkpointing=False, tiny_llama=False):
+    def adjust_learning_rate(param_group, LR, epoch):
+        # Decay the learning rate with half-cycle cosine after warmup
```

Comment:
Any reason or reference for this LR adjustments?

Reply:
original authors of g-retriever paper from meta use this